### PR TITLE
Catch JS errors and send them to the API

### PR DESCRIPTION
**Pull request by scruffian**
I made some more changes to this to remove the file from the Webpack build. Initially this was to avoid an IE11 bug, but I actually think this is a better approach, as this will also catch errors with the Webpack build.

As @blowery pointed out, we will be losing errors to the debouncer, so we'll need to go back and address that at some point.

To make sure you are included in logging add this to your local storage:
`localStorage.setItem('log-errors', true);`

Also you will need to build calypso as a different environment and ensure you remove this line:
https://github.com/Automattic/wp-calypso/pull/3089/files#diff-1a680164864b98b0fe2cbb81bea45dc8R77

To build using the wpcalypso environment:

`CALYPSO_ENV=wpcalypso make run`

This is necessary as the minified version only gets built when we build the bundles for a non dev environment because it happens in the bundler.

### CODE REVIEW

**Comment by blowery** [comment]
I had a thought on an alternative to debouncing. Our onerror handler could always push the error into an array and then set a timer to drain the error queue. If the timer is already set, don't set another one.

Then the drain can unset the timer handle and push out all of the errors in one API call (or a batch if there are too many) and reset the queue. 

Our debouncer is currently set to 100ms, which seems a bit low. I'd think we'd want something more like 1000ms? Potentially 10 requests a second seems pretty high.

**Comment by stephanethomas** [comment]
Tested and got the following line in our administration tool:

```
(5:58:32 PM) calypso-js-error: {"stack":"Error: Script error.
```

The original error was:

screenshot

Was it the expected behavior? I might have missed something but right now there the output is not very useful :(.

**Comment by Tug** [comment]
Yep, that's the issue with `window.onerror`, the error information is often lost to a `Script Error.`

**Comment by scruffian** [comment]
@blowery I added a commit to save the errors to an array, and then a timer to send them to the API every 10 seconds. What do you think?

File: catch-js-errors.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		// POST to the API
+		xhr.open( 'POST', 'https://public-api.wordpress.com/rest/v1.1/js-error?http_envelope=1', true );
+		xhr.setRequestHeader( 'Content-type', 'application/x-www-form-urlencoded' );
+
+		params = 'client_id=39911&client_secret=cOaYKdrkgXz8xY7aysv4fU6wL6sK5J8a6ojReEIAPwggsznj4Cb6mW0nffTxtYT8&error=' + stringifyErrorForUrlParams( error );
+		xhr.send( params );
+	}
+
+	if ( isLocalStorageNameSupported() ) {
+		// Randomly assign 1% of users to log errors
```
Comment (Tug) [inline catch-js-errors.js]: 1% of horizon/wpcalypso users, isn't that a bit low?

**Comment by Tug** [comment]
@scruffian Did you forget to push the change?

**Comment by scruffian** [comment]
> @scruffian Did you forget to push the change?

I did, sorry

**Comment by scruffian** [comment]
Thanks for the suggestions @Tug, I included them in  3a46e53.

**Comment by Tug** [comment]
LGTM, let's try it in staging :+1: 
Could we remove the random 1% assignment while we're testing?

**Pull request merged**